sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala

```diff
@@ -853,4 +854,32 @@ class JsonExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper with
         checkEvaluation(JsonObjectKeys(Literal(input)), expected)
     }
   }
+
+  def checkMapTypeKeyError(schema: DataType, jsonData: String): Unit = {
+    assert( JsonToStructs(schema, Map.empty, Literal(jsonData)).checkInputDataTypes() match {
+      case TypeCheckFailure(_) => true
+      case _ => false
+    })
+  }
+
+  test("SPARK-35320: from_json should fail with a key type different of StringType") {
+    checkMapTypeKeyError(MapType(IntegerType, StringType), """{"1": "test"}""")
+  }
+
+  test("SPARK-35320: from_json should fail with a key type different of StringType " +
+    "anidated in a StructType") {
+    checkMapTypeKeyError(StructType(Seq(StructField("test", MapType(IntegerType, StringType)))),
+      """"test": {"1": "test"}""")
+  }
+
+  test("SPARK-35320: from_json should fail with a key type different of StringType " +
+    "anidated in a ArrayType") {
+    checkMapTypeKeyError(ArrayType(MapType(IntegerType, StringType)), """[{"1": "test"}]""")
+  }
+
+  test("SPARK-35320: from_json should fail with a key type different of StringType " +
+    "anidated in a MapType") {
+    checkMapTypeKeyError(MapType(StringType, MapType(IntegerType, StringType)),
+    """{"key": {"1" : "test"}}""")
+  }
```

Comment:
Looks like you can have just one test and run through the combination of map types and json string.

Reply:
Yes you are right, I'm going to put it in only one test to simplify it